umap/umap_.py

```diff
@@ -1688,8 +1689,43 @@ def fit(self, X, y=None):
         if self.verbose:
             print("Construct fuzzy simplicial set")
 
+        if self.metric == 'precomputed' and self._sparse_data:
+            # For sparse precomputed distance matrices, we just argsort the rows to find nearest neighbors
+            # To make this easier, we expect matrices that are symmetrical (so we can find neighbors by looking at
+            # rows in isolation, rather than also having to consider that sample's column too).
+            print("Computing KNNs for sparse precomputed distances...")
+            if sparse_tril(X).getnnz() != sparse_triu(X).getnnz():
+                raise ValueError("Sparse precomputed distance matrices should be symmetrical!")
+            self._knn_indices = np.zeros((X.shape[0], self.n_neighbors), dtype=np.int)
+            self._knn_dists = np.zeros(self._knn_indices.shape, dtype=np.float)
+            for row_id in range(X.shape[0]):
+                # Find KNNs row-by-row
+                row_data = X[row_id].data
+                row_indices = X[row_id].indices
+                # Remove distance from sample to itself (i.e. the matrix diagonal), if present
+                row_data = row_data[row_indices != row_id]
+                row_indices = row_indices[row_indices != row_id]
```

Comment:
Usually the knn includes the point itself, and distances of zero are handled specially. I suspect this doesn't break anything, but I believe it is not necessary either (unless samples potentially have non-zero distances to themselves).

Reply:
I removed these and replaced them with a check that all the entries on the diagonal are either missing or zero.

Reply:
Do you want me to explicitly insert zero values on the diagonal, to ensure consistency?